addon/components/discover-page/component.js

```diff
@@ -0,0 +1,808 @@
+                let result = Ember.merge(hit._source, {
+                    id: hit._id,
+                    type: 'elastic-search-result',
+                    workType: hit._source['@type'],
+                    abstract: hit._source.description,
+                    subjects: hit._source.subjects.map(each => ({ text: each })),
+                    providers: hit._source.sources.map(item => ({ name: item })), // For PREPRINTS, REGISTRIES
+                    hyperLinks: [// Links that are hyperlinks from hit._source.lists.links
+                        {
+                            type: 'share',
+                            url: config.OSF.shareBaseUrl + `${hit._source.type}` + '/' + hit._id
+                        }
+                    ],
+                    infoLinks: [], // Links that are not hyperlinks  hit._source.lists.links
+                    registrationType: hit._source.registration_type // For REGISTRIES
+                });
+
+                hit._source.identifiers.forEach(function(identifier) {
+                    if (identifier.startsWith('http://')) {
+                        result.hyperLinks.push({ url: identifier });
+                    } else {
+                        const spl = identifier.split('://');
+                        const [type, uri, ..._] = spl; // jshint ignore:line
+                        result.infoLinks.push({ type, uri });
+                    }
+                });
+
+                result.contributors = result.lists.contributors ? result.lists.contributors
+                  .sort((b, a) => (b.order_cited || -1) - (a.order_cited || -1))
+                  .map(contributor => ({
+                        users: Object.keys(contributor)
+                          .reduce(
+                              (acc, key) => Ember.merge(acc, { [Ember.String.camelize(key)]: contributor[key] }),
+                              { bibliographic: contributor.relation !== 'contributor' }
+                          )
+                    })) : [];
+
+                // Temporary fix to handle half way migrated SHARE ES
+                // Only false will result in a false here.
+                result.contributors.map(contributor => contributor.users.bibliographic = !(contributor.users.bibliographic === false));  // jshint ignore:line
+
+                return result;
+            });
+
+            if (json.aggregations) {
+                this.set('aggregations', json.aggregations);
+            }
+            this.setProperties({
+                numberOfResults: json.hits.total,
+                took: moment.duration(json.took).asSeconds(),
+                loading: false,
+                firstLoad: false,
+                results: results,
+                queryError: false,
+                shareDown: false,
+            });
+            if (this.get('totalPages') && this.get('totalPages') < this.get('page')) {
+                this.search();
+            }
+        }, (errorResponse) => {
```

Comment:
That's a good point- deferreds are slightly different, and .fail would probably work. The main goal is to catch errors that happen inside the `.then` block.

To help us treat all requests via the same promise semantics, I typically do something like the following; we really should abstract this to a helper function at some point. (especially as we add more code to make manual AJAX requests- I'd much prefer one consistent API for interacting with all promises!)

```
        let jqDeferred = this._doRequest();
        return new Ember.RSVP.Promise((resolve, reject) => {
            jqDeferred.done((value) => resolve(value));
            jqDeferred.fail((reason) => reject(reason));
        });
```

Addons like `ember-ajax` do something [similar](https://github.com/ember-cli/ember-ajax/blob/master/addon/mixins/ajax-request.js#L212-L218).